modules/security/src/test/java/org/apache/ignite/internal/processor/security/client/ThinClientPermissionCheckTest.java

```diff
@@ -0,0 +1,266 @@
+            executeOperation(CLIENT, c);
+
+        for (Consumer<IgniteClient> c : consumers(FORBIDDEN_CACHE))
+            executeForbiddenOperation(c);
+    }
+
+    /**
+     * That test shows the wrong case when a client has permission for a remove operation
+     * but a removeAll operation is forbidden for it. To have permission for the removeAll (clear) operation
+     * a client need to have the permission to execute {@link #REMOVE_ALL_TASK} ({@link #CLEAR_TASK}) task.
+     *
+     * @throws Exception If error occurs.
+     */
+    @Test
+    public void testCacheTaskPermOperations() throws Exception {
+        executeOperation(CLIENT_CACHE_TASK_OPER, c -> c.cache(CACHE).removeAll());
+        executeOperation(CLIENT_CACHE_TASK_OPER, c -> c.cache(CACHE).clear());
+
+        executeForbiddenOperation(c -> c.cache(CACHE).removeAll());
+        executeForbiddenOperation(c -> c.cache(CACHE).clear());
+    }
+
+    /**
+     * @throws Exception If error occurs.
+     */
+    @Test
+    public void testSysOperation() throws Exception {
+        try (IgniteClient sysPrmClnt = startClient(CLIENT_SYS_PERM)) {
+            sysPrmClnt.createCache(DYNAMIC_CACHE);
+
+            assertThat(sysPrmClnt.cacheNames().contains(DYNAMIC_CACHE), is(true));
+
+            sysPrmClnt.destroyCache(DYNAMIC_CACHE);
+
+            assertThat(sysPrmClnt.cacheNames().contains(DYNAMIC_CACHE), is(false));
+        }
+
+        executeForbiddenOperation(c -> c.createCache(DYNAMIC_CACHE));
+        executeForbiddenOperation(c -> c.destroyCache(CACHE));
+    }
+
+    /**
+     * @param cacheName Cache name.
+     */
+    private Collection<Consumer<IgniteClient>> consumers(final String cacheName) {
+        return Arrays.asList(
+            c -> c.cache(cacheName).put("key", "value"),
+            c -> c.cache(cacheName).putAll(singletonMap("key", "value")),
+            c -> c.cache(cacheName).get("key"),
+            c -> c.cache(cacheName).getAll(Collections.singleton("key")),
+            c -> c.cache(cacheName).containsKey("key"),
+            c -> c.cache(cacheName).remove("key"),
+            c -> c.cache(cacheName).replace("key", "value"),
+            c -> c.cache(cacheName).putIfAbsent("key", "value"),
+            c -> c.cache(cacheName).getAndPut("key", "value"),
+            c -> c.cache(cacheName).getAndRemove("key"),
+            c -> c.cache(cacheName).getAndReplace("key", "value")
+        );
+    }
+
+    /**
+     * @param cons Consumer.
+     */
+    private void executeOperation(String clientName, Consumer<IgniteClient> cons) throws Exception {
+        try (IgniteClient client = startClient(clientName)) {
+            cons.accept(client);
+        }
+    }
+
+    /**
+     * @param cons Consumer.
+     */
+    private void executeForbiddenOperation(Consumer<IgniteClient> cons) {
+        try (IgniteClient client = startClient(CLIENT)) {
+            cons.accept(client);
+
+            fail();
```

Comment:
Let's write here "This operation has to be forbidden". Ideally, we should write an operation name in the exception message.